ksqldb-execution/src/main/java/io/confluent/ksql/execution/interpreter/ExpressionInterpreter.java

```diff
@@ -0,0 +1,769 @@
+      switch (node.getSign()) {
+        case MINUS:
+          return visitArithmeticMinus(value);
+        case PLUS:
+          return visitArithmeticPlus(value);
+        default:
+          throw new UnsupportedOperationException("Unsupported sign: " + node.getSign());
+      }
+    }
+
+    private Pair<Object, SqlType> visitArithmeticMinus(final Pair<Object, SqlType> value) {
+      if (value.getLeft() instanceof BigDecimal) {
+        final BigDecimal bigDecimal = (BigDecimal) value.getLeft();
+        return new Pair<>(
+            bigDecimal.negate(new MathContext(((SqlDecimal) value.getRight()).getPrecision(),
+                RoundingMode.UNNECESSARY)),
+            value.getRight()
+        );
+      } else if (value.getRight().baseType() == SqlBaseType.DOUBLE) {
+        final double val = (Double) value.getLeft();
+        return new Pair<>(-val, value.getRight());
+      } else if (value.getRight().baseType() == SqlBaseType.INTEGER) {
+        final int val = (Integer) value.getLeft();
+        return new Pair<>(-val, value.getRight());
+      } else if (value.getRight().baseType() == SqlBaseType.BIGINT) {
+        final long val = (Long) value.getLeft();
+        return new Pair<>(-val, value.getRight());
+      } else {
+        throw new UnsupportedOperationException("Negation on unsupported type: "
+            + value.getLeft());
+      }
+    }
+
+    private Pair<Object, SqlType> visitArithmeticPlus(final Pair<Object, SqlType> value) {
+      if (value.getRight().baseType() == SqlBaseType.DECIMAL) {
+        final BigDecimal bigDecimal = (BigDecimal) value.getLeft();
+        return new Pair<>(
+            bigDecimal.plus(new MathContext(((SqlDecimal) value.getRight()).getPrecision(),
+                RoundingMode.UNNECESSARY)),
+            value.getRight()
+        );
+      } else if (value.getRight().baseType() == SqlBaseType.DOUBLE) {
+        final double val = (Double) value.getLeft();
+        return new Pair<>(+val, value.getRight());
+      } else if (value.getRight().baseType() == SqlBaseType.INTEGER) {
+        final int val = (Integer) value.getLeft();
+        return new Pair<>(+val, value.getRight());
+      } else if (value.getRight().baseType() == SqlBaseType.BIGINT) {
+        final long val = (Long) value.getLeft();
+        return new Pair<>(+val, value.getRight());
+      } else {
+        throw new UnsupportedOperationException("Unary plus on unsupported type: "
+            + value.getLeft());
+      }
+    }
+
+    @Override
+    public Pair<Object, SqlType> visitArithmeticBinary(
+        final ArithmeticBinaryExpression node, final Void context
+    ) {
+      final Pair<Object, SqlType> left = process(node.getLeft(), context);
+      final Pair<Object, SqlType> right = process(node.getRight(), context);
+
+      final SqlType schema = expressionTypeManager.getExpressionSqlType(node);
+
+      if (schema.baseType() == SqlBaseType.DECIMAL) {
```

Comment:
Why not put the `if` condition inside `doArithmetic` as well? Then all `apply` functions within the other class can be private right?